File: cukcuk clone/app/src/main/java/vn/com/misa/cukcukstarterclone/data/repository/UserRepository.java
package vn.com.misa.cukcukstarterclone.data.repository;

import vn.com.misa.cukcukstarterclone.data.IOnLoadedCallback;
import vn.com.misa.cukcukstarterclone.data.source.IUserDataSource;

/**
 * @created_by KhanhNQ on 08-Feb-2021.
 */
public class UserRepository implements IUserRepository {

    private final IUserDataSource.Remote remoteDataSource;

    private UserRepository(IUserDataSource.Remote userRemoteDataSource) {
        this.remoteDataSource = userRemoteDataSource;
    }

    private static UserRepository instance;

    public static UserRepository getInstance(IUserDataSource.Remote userRemoteDataSource) {
        if (instance == null) {
            instance = new UserRepository(userRemoteDataSource);
        }
        return instance;
    }

    @Override
    public void login(String username, String password, IOnLoadedCallback<Boolean> callback) {
        remoteDataSource.login(username, password, callback);
    }

    @Override
    public void register(String username, String password, IOnLoadedCallback<Boolean> callback) {
        remoteDataSource.register(username, password, callback);
    }
}

File: app/src/main/java/com/misa/fresher/data/repository/user/UserRepository.kt
package com.misa.fresher.data.repository.user

import com.misa.fresher.data.IOnLoadedCallback
import com.misa.fresher.data.model.user.UserResponse
import com.misa.fresher.data.model.user.UserError
import com.misa.fresher.data.source.IUserDataSource


/**
 * - class's purpose:
 *
 * @author HTLong
 * @edit_at 2022-03-21
 */
class UserRepository private constructor(private val userRemoteDataSource: IUserDataSource.Remote) : IUserRepository {
    override fun signIn(username: String, password: String, callback: IOnLoadedCallback<UserResponse, UserError>) {
        userRemoteDataSource.signIn(username, password, callback)
    }

    override fun signUp(username: String, password: String, callback: IOnLoadedCallback<UserResponse, UserError>) {
        userRemoteDataSource.signUp(username, password, callback)
    }

    companion object {
        private var instance: UserRepository? = null
        fun getInstance(userRemoteDataSource: IUserDataSource.Remote) = instance ?: UserRepository(userRemoteDataSource).also { instance = it }
    }
}
